baseline-error-prone/src/main/java/com/palantir/baseline/errorprone/LambdaMethodReference.java

```diff
@@ -0,0 +1,123 @@
+/*
+ * (c) Copyright 2019 Palantir Technologies Inc. All rights reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.palantir.baseline.errorprone;
+
+import com.google.auto.service.AutoService;
+import com.google.errorprone.BugPattern;
+import com.google.errorprone.VisitorState;
+import com.google.errorprone.bugpatterns.BugChecker;
+import com.google.errorprone.fixes.SuggestedFix;
+import com.google.errorprone.fixes.SuggestedFixes;
+import com.google.errorprone.matchers.Description;
+import com.google.errorprone.util.ASTHelpers;
+import com.sun.source.tree.BlockTree;
+import com.sun.source.tree.ExpressionTree;
+import com.sun.source.tree.LambdaExpressionTree;
+import com.sun.source.tree.MethodInvocationTree;
+import com.sun.source.tree.ReturnTree;
+import com.sun.source.tree.StatementTree;
+import com.sun.source.tree.Tree;
+import com.sun.tools.javac.code.Symbol;
+import java.util.Optional;
+
+@AutoService(BugChecker.class)
+@BugPattern(
+        name = "LambdaMethodReference",
+        link = "https://github.com/palantir/gradle-baseline#baseline-error-prone-checks",
+        linkType = BugPattern.LinkType.CUSTOM,
+        severity = BugPattern.SeverityLevel.SUGGESTION,
+        summary = "Lambda should be a method reference")
+public final class LambdaMethodReference extends BugChecker implements BugChecker.LambdaExpressionTreeMatcher {
+
+    private static final String MESSAGE = "Lambda should be a method reference";
+
+    @Override
+    public Description matchLambdaExpression(LambdaExpressionTree tree, VisitorState state) {
+        // Only handle simple no-arg method references for the time being, don't worry about
+        // simplifying map.forEach((k, v) -> func(k, v)) to map.forEach(this::func)
+        if (!tree.getParameters().isEmpty()) {
+            return Description.NO_MATCH;
+        }
+        LambdaExpressionTree.BodyKind bodyKind = tree.getBodyKind();
+        Tree body = tree.getBody();
+        // n.b. These checks are meant to avoid any and all cleverness. The goal is to be confident
+        // that we can fix the most basic problems correctly, not to take risks and produce code
+        // that may not compile.
+        switch (bodyKind) {
```

Comment:
could we use a visitor instead of reflection here?
```suggestion
body.accept(new SimpleTreeVisitor<Description, Object>(Description.NO_MATCH) {
	@Override
    public Description visitMethodInvocation(MethodInvocationTree node, Object o) {
    	...
    }
    
    @Override
    public Description visitBlock(BlockTree node, Object o) {
		...
    }
}, null)
```

Reply:
It would be cleaner, but I'm worried that this check will run a lot. We've had issues with errorprone performance in the past, so I avoided allocation until we've discovered a failure.

Reply:
I'm not opposed to trying it, and walking back if we see degraded build times